### PR TITLE
Fix large datasets not saving

### PR DESCRIPTION
**Pull request by charlesbaynham**
# ARTIQ Pull Request

## Description of Changes

Large datasets were not being saved properly: see https://forum.m-labs.hk/d/519-how-large-can-a-dataset-be-before-there-is-a-problem-with-saving-hdf5/11 for an exploration. 

This was due to a race condition in

https://github.com/m-labs/artiq/blob/58cc3b8d0a957b24834e08cb3fbe70c148fd17b4/artiq/master/worker_impl.py#L347-L356

...where the worker was reporting that it had finished before it had actually started saving HDF5 files. It then had two seconds before the timeout in 

https://github.com/m-labs/artiq/blob/58cc3b8d0a957b24834e08cb3fbe70c148fd17b4/artiq/master/worker.py#L123-L134

...killed it. 

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :bug: Bug fix  |
|   | :sparkles: New feature |
|   | :hammer: Refactoring  |
|   | :scroll: Docs |

### All Pull Requests

- [x] Use correct spelling and grammar.
- [ ] Update [RELEASE_NOTES.rst](../RELEASE_NOTES.rst) if there are noteworthy changes, especially if there are changes to existing APIs.
- [ ] Close/update issues.
- [ ] Check the copyright situation of your changes and sign off your patches (`git commit --signoff`, see [copyright](../CONTRIBUTING.rst#copyright-and-sign-off)).

### Code Changes

- [ ] Run `flake8` to check code style (follow PEP-8 style). `flake8` has issues with parsing Migen/gateware code, ignore as necessary.
- [x] Test your changes or have someone test them. Mention what was tested and how.
- [ ] Add and check docstrings and comments
- [ ] Check, test, and update the [unittests in /artiq/test/](../artiq/test/) or [gateware simulations in /artiq/gateware/test](../artiq/gateware/test)

### Documentation Changes

- [ ] Check, test, and update the documentation in [doc/](../doc/). Build documentation (`cd doc/manual/; make html`) to ensure no errors.

### Git Logistics

- [ ] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [ ] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

### CODE REVIEW

File: artiq/master/worker_impl.py [hunk 1]
```diff
@@ -347,12 +347,13 @@ def write_results():
...
                 try:
                     exp_inst.analyze()
-                    put_completed()
                 finally:
                     # browser's analyze shouldn't write results,
                     # since it doesn't run the experiment and cannot have rid
                     if rid is not None:
                         write_results()
+
+                    put_completed()
```
[the remark this answers is not in the document]
Comment (dnadlinger) [reply artiq/master/worker_impl.py:356]: (It seems like there is a hole in experiment lifecycle test coverage then – either that, or I misunderstand what's going on here.)

[the remark this answers is not in the document]
Comment (charlesbaynham) [reply artiq/master/worker_impl.py:356]: Yes, you're right, that's better - I've pushed an update

**Comment by dnadlinger** [comment]
Could you roll back the formatting changes, please? I'll merge it then.

**Comment by charlesbaynham** [comment]
Oops, those snuck in. Reversed. 

+1 for mandating black everywhere. If we drop in a `pre-commit-config.yaml` then we can have https://pre-commit.ci/ push suggested fixes to PRs automatically. I can make a PR if people approve, though it needs a maintainer to enable the PR fixer.

**Review by dnadlinger — APPROVED** [review]

**Comment by charlesbaynham** [comment]
> > +1 for mandating black everywhere. If we drop in a `pre-commit-config.yaml` then we can have https://pre-commit.ci/ push suggested fixes to PRs automatically. I can make a PR if people approve, though it needs a maintainer to enable the PR fixer.
> 
> I'd rather have something that doesn't involve cloud/SaaS/proprietary solutions, e.g. a bot that can run on our own server and supports both GitHub and others would be much better.

Pre commit runs locally and is open source, it's just the bot that's in the cloud. Putting it into a gitlab / hydra pipeline is easy, the bot just adds integration with Github's PR API.

**Pull request merged**